### PR TITLE
Publish only the match a backward search answers with

### PR DESCRIPTION
**Pull request by takumin**
Stacked on #7148, which this needs and whose commits it carries. Review the last two.

#7148 took the character counting out of the walk `String#rindex`, `String#byterindex` and `String#rpartition` share, and named what it did not fix: every search publishes its match, cutting the whole subject into the pre match and post match globals, so the walk still copies the subject once per match. This is that.

```
                                    #7148     here
("あb" *  5000).rindex(/b/)         0.007s    0.002s
("あb" * 10000).rindex(/b/)         0.019s    0.004s
("あb" * 20000).rindex(/b/)         0.060s    0.007s
("あb" * 10000).byterindex(/b/)     0.017s    0.004s
("あb" * 10000).rpartition(/b/)     0.016s    0.004s
("ab"  * 10000).rindex(/b/)         0.012s    0.004s
("ab"  * 20000).rindex(/b/)         0.034s    0.007s
```

Doubling the subject now doubles the time rather than tripling it. The single-byte subject was paying the same copy and had nothing to do with characters, which is why it moves too.

### What was being thrown away

A successful search publishes thirteen names: `$~`, `$1` through `$9`, the match, the two pieces the subject is cut into around it, and the last group that took part. Two of those thirteen are cut from the subject and together are the whole of it, so a search costs its subject once over whether or not anything reads the result.

The walk keeps the last match that qualifies, so every match it passes is published and then replaced by the next one. It already publishes its answer itself at the end, with `MatchData#__set_globals`, and clears for a miss. Nothing read what the searches published in between.

### Letting a search publish nothing

`Regexp.__byte_search` takes it as an argument, the way CRuby's `rb_reg_search0()` takes `set_backref_str`, and it reaches `exec_match()` and `create_matchdata()` from there. A search that publishes nothing clears nothing either, so the globals come out of it exactly as they went in and the caller owns them throughout; the walk passes FALSE and keeps the two lines that publish its answer or clear for a miss.

`gsub`, `scan` and `sub` cannot be told the same thing: the block they call reads the globals of the match it was handed, so those loops go on publishing every turn. The argument defaults to TRUE, so every other caller is where it was.

### Tests

The first commit asks for the eleven globals no test read back. `$~` and `$1` after a backward search are asserted already; the match, the two pieces, a second group, a group that did not take part, the last one that did, and the same set after a miss were not, and they are exactly what a walk that passes matches on the way to its answer has to get right. Every answer is CRuby's.

The second commit is where they start being published by one act rather than falling out of the last search. Keeping the walk's own publish but dropping its clear, which is the mistake this change makes available now that a failing search no longer clears, turns 6 of the new assertions red.

The differential from #7148 was rerun against this: 21,672 checks on the UTF-8 build and 24,552 on the byte-indexing one, comparing the answer, the exception where one is raised, and `$~`, `$&`, `` $` ``, `$'` and `$1` after every call. 0 mismatches.

### Verification

Full suite green at every commit on `build_config/ci/gcc-clang.rb`: full-core with `MRB_GC_STRESS` (2282), bintest (2283), the C++ ABI (2283), and the default gembox (2068). 0 failures, 0 crashes, no new compiler warnings.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved reverse regular-expression searches for multibyte text.
  * Corrected character-based and byte-based index limits, including negative positions, overlapping matches, and partitioning.
  * Ensured regular-expression match results are consistently updated after successful searches and cleared when searches fail.

* **Tests**
  * Added coverage for multibyte reverse searches, boundary conditions, overlapping matches, partitioning, and match-result handling.

### CODE REVIEW

File: mrbgems/mruby-regexp/test/string_index.rb [hunk 1]
```diff
@@ -544,6 +578,45 @@ def fake.match(str, pos = 0); raise "must not be called"; end
...
+  assert_equal "b", $+
+
+  # and a search that finds nothing clears all of them, not `$~` alone
+  "zzz" =~ /(z)/
+  assert_nil "abc".rindex(/x/)
+  assert_nil $&
+  assert_nil $`
+  assert_nil $'
+  assert_nil $1
+  assert_nil $+
```
[the remark this answers is not in the document]
Comment (takumin) [reply mrbgems/mruby-regexp/test/string_index.rb:617]: Done in #7151. Seeded two groups and added `assert_nil $~` and `assert_nil $2`: `$2` was nil before the call as you say, so the assertion stood for nothing until the seed filled it.
